core/src/audio_subsystem.cpp

```diff
@@ -0,0 +1,206 @@
+
+            if (m_result == FMOD_ERR_INVALID_HANDLE) {
+                source.set_channel(nullptr);
+                continue;
+            }
+
+            m_result = source.get_channel()->getPaused(&paused);
+            ErrorCheck(m_result);
+            m_result = source.get_channel()->getPosition(&pos, FMOD_TIMEUNIT_MS);
+            ErrorCheck(m_result);
+            source.get_channel()->getCurrentSound(&current_sound);
+
+            if (current_sound) {
+                m_result = current_sound->getLength(&len_ms, FMOD_TIMEUNIT_MS);
+                ErrorCheck(m_result);
+            }
+        }
+    }
+}
+
+void AudioSubsystem::set_loop(components::AudioSource& source, bool loops) {
+    source.set_loops(loops);
+
+    m_result = source.get_sound()->setMode(source.get_loops() ? FMOD_LOOP_NORMAL : FMOD_LOOP_OFF);
+    ErrorCheck(m_result);
+}
+
+void AudioSubsystem::add_sound(components::AudioSource& source) {
+    log::info("TRYING TO LOAD AUDIO FILE {}", source.get_path().string());
+    m_result =
+        m_system->createSound(source.get_path().string().c_str(), FMOD_3D_LINEARROLLOFF, nullptr, &source.get_sound());
+    ErrorCheck(m_result);
+
+    m_result = source.get_sound()->set3DMinMaxDistance(0.0f, 50000.0f);
+    ErrorCheck(m_result);
+    set_loop(source, source.get_loops());
+    source.get_sound()->setDefaults(12000, 128);
+}
+
+void AudioSubsystem::on_update(double m_delta_time) {
+    auto sceneOpt = Scene::get_active_scene();
+    if (!sceneOpt) {
+        return;
+    }
+    Scene& scene = sceneOpt.value();
+    entt::registry& registry = scene.get_registry();
+
+    auto sources = registry.view<components::AudioSource>();
+
+    for (auto sourceHandle : sources) {
+        auto sourceOpt = scene.get_game_object_from_handle(sourceHandle);
+        if (!sourceOpt) {
+            continue;
+        }
+        auto sourceGO = sourceOpt.value();
+
+        auto& source = sourceGO.get_component<components::AudioSource>();
+        this->update_source(source);
+    }
+
+    auto listeners = registry.view<components::AudioListener>();
+    for (auto listenerHandle : listeners) {
+        auto listenerOpt = scene.get_game_object_from_handle(listenerHandle);
+        if (!listenerOpt) {
+            continue;
+        }
+        auto listenerGO = listenerOpt.value();
+
+        auto& listener = listenerGO.get_component<components::AudioListener>();
+        this->update_listener(listener);
+    }
+}
+
+void AudioSubsystem::update_source(components::AudioSource& source) {
+    FMOD_VECTOR* pos = source.get_position();
+    FMOD_VECTOR vel = {0, 0, 0};
+
+    if (source.get_channel()) {
+        m_result = source.get_channel()->set3DAttributes(pos, &vel);
+        ErrorCheck(m_result);
+    }
+
+    update();
+}
+
+void AudioSubsystem::update_listener(components::AudioListener& listener) {
+    FMOD_VECTOR forward = {listener.get_forward()->x, listener.get_forward()->y, listener.get_forward()->z};
```

Comment:
Don't know if `get_forwards` calculates anything, but if it does you are making it calculate three times, when you can save it to a `vec3` variable and then pass those values here